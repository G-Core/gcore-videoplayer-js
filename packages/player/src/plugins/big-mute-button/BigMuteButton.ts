import { Events, template, UICorePlugin, Utils } from '@clappr/core'
import { trace } from '@gcorevideo/utils'

import { CLAPPR_VERSION } from '../../build.js'

import volumeMuteIcon from '../../../assets/icons/new/volume-off.svg'
import templateHtml from '../../../assets/big-mute-button/big-mute-button.ejs'
import '../../../assets/big-mute-button/big-mute-button.scss'

const T = 'plugins.big_mute_button'

// TODO rewrite as a container plugin

/**
 * `PLUGIN` that displays a big mute button over the video when it's being played muted.
 * @public
 * @remarks
 * When pressed, it unmutes the video.
 * @example
 * ```ts
 * import { BigMuteButton } from '@gcorevideo/player'
 * Player.registerPlugin(BigMuteButton)
 * ```
 */
export class BigMuteButton extends UICorePlugin {
  private autoPlay = false

  private hidden = false

  // TODO get back to the ads-related logic later
  private _adIsPlaying = false

  /**
   * @internal
   */
  get name() {
    return 'big_mute_button'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  private static readonly template = template(templateHtml)

  /**
   * @internal
   */
  override get events() {
    return {
      click: 'clicked',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenTo(this.core, Events.CORE_READY, this.onCoreReady)
    this.listenTo(
      this.core,
      Events.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onContainerChanged,
    )
    this.listenTo(this.core, 'core:advertisement:start', this.onStartAd)
    this.listenTo(this.core, 'core:advertisement:finish', this.onFinishAd)
  }

  private onCoreReady() {}

  private onContainerChanged() {
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_VOLUME,
      this.onContainerVolume,
    )
    this.listenTo(
      this.core.activePlayback,
      Events.PLAYBACK_ENDED,
      this.onPlaybackEnded,
    )
    this.listenTo(this.core.activeContainer, Events.CONTAINER_PLAY, this.onPlay)
    this.listenTo(this.core.activeContainer, Events.CONTAINER_STOP, this.onStop)
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_PAUSE,
      this.onPause,
    )
  }

  private onPlay(_: string, { autoPlay }: { autoPlay?: boolean }) {
    const container = this.core.activeContainer
    const { volume } = container
    const { wasMuted } = this.options
    if (autoPlay) {
      this.autoPlay = true
    }
    trace(`${T} onPlay`, {
      autoPlay: this.autoPlay,
      wasMuted,
      volume,
    })
    if (this.autoPlay && !wasMuted && volume === 0) {
      this.mount()
    } else {
      this.destroy()
    }
  }

  private onStop(_: string, metadata?: { ui?: boolean }) {
    const ui = metadata?.ui
    trace(`${T} onStop`, { ui })
    if (ui) {
      this.destroy()
    }
  }

  private onPause() {
    this.destroy()
  }

  private onContainerVolume(value: number) {
    if (value !== 0) {
      this.destroy()
    }
  }

  private onPlaybackEnded() {
    this.hide()
  }

  private onStartAd() {
    this._adIsPlaying = true
    this.hide()
  }

  private onFinishAd() {
    this._adIsPlaying = false
    this.show()
  }

  /**
   * @internal
   */
  override render() {
    this.$el.html(BigMuteButton.template())
    this.$el.find('#gplayer-big-mute-icon').append(volumeMuteIcon)

    // TODO
    // this._adIsPlaying && this.hide()

    return this
  }

  private mount() {
    this.core.activeContainer.$el.append(this.$el)
    this.show()
  }

  private hide() {
    this.hidden = true
    this.$el.find('#gplayer-big-mute-button')?.addClass('hide')
  }

  private show() {
    this.hidden = false
    this.$el.find('#gplayer-big-mute-button')?.removeClass('hide')
  }

  private clicked(e: MouseEvent) {
    const mediaControl = this.core.getPlugin('media_control')
    // TODO delegate to media_control plugin
    const localVolume = Utils.Config.restore('volume')
    const volume = !isNaN(localVolume) ? localVolume : 100
    const unmuted = volume === 0 ? 100 : volume

    if (mediaControl) {
      mediaControl.setVolume(unmuted)
    } else {
      this.core.activeContainer.setVolume(unmuted)
    }

    e.stopPropagation?.()

    this.destroy()
  }
}
