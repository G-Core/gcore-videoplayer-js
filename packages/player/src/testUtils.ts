import { $, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(options),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    isAudioOnly: false,
    levels: [],
    consent: vi.fn(),
    play: vi.fn(),
    pause: vi.fn(),
    stop: vi.fn(),
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(
  options: Record<string, unknown> = {},
  playback: any = createMockPlayback(),
) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    options: {
      ...options,
    },
    $el: $(el),
    disableMediaControl: vi.fn(),
    enableMediaControl: vi.fn(),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockReturnValue(0),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isHighDefinitionInUse: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  // TODO <div class="media-control-layer">
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  // @ts-ignore
  mediaControl.mount = vi.fn()
  // @ts-ignore
  mediaControl.container = core.activeContainer
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  vi.spyOn(mediaControl, 'trigger')
  core.$el.append(mediaControl.$el)
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>')
      .attr(`data-${name}`, '')
      .append($el)
      .appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
