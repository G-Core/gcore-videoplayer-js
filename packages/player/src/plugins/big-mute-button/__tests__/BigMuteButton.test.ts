import { beforeEach, describe, it, expect } from 'vitest'
import { Events } from '@clappr/core'

import { BigMuteButton } from '../BigMuteButton.js'
import { createMockCore } from '../../../testUtils.js'

describe('BigMuteButton', () => {
  let core: any
  let bmb: BigMuteButton
  describe('basically', () => {
    beforeEach(() => {
      core = createMockCore({})
      bmb = new BigMuteButton(core)
      // core.emit('core:ready')
      // core.emit('core:active:container:changed')
    })
    it('should render', () => {
      expect(bmb.$el.html()).toMatchSnapshot()
    })
  })
  describe('when container starts playing', () => {
    describe.each([
      ['muted autoplay', 0, { autoPlay: true }, true],
      ['audible autoplay', 50, { autoPlay: true }, false],
      ['muted not autoplay', 0, { }, false],
      ['audible not autoplay', 1, {}, false],
    ])("%s", (_, volume, playMetadata, shouldMount) => {
      beforeEach(() => {
        core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED)
        core.activeContainer.volume = volume
        core.activeContainer.emit(Events.CONTAINER_PLAY, 'Container', playMetadata)
      })
      it(`should ${shouldMount ? 'mount' : 'not mount'} to container`, () => {
        expect(core.activeContainer.$el.find('#gplayer-big-mute-button').length).toBe(shouldMount ? 1 : 0)
      })
    })
  })
})