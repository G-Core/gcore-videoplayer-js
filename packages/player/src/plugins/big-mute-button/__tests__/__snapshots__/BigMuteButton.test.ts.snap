// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BigMuteButton > basically > should render 1`] = `
"<div class="big-mute-icon-wrapper" data-big-mute="" id="gplayer-big-mute-button">
    <div class="big-mute-icon gcore-skin-border-color" data-big-mute-icon="" id="gplayer-big-mute-icon">/assets/icons/new/volume-off.svg</div>
</div>
"
`;
